feat: ooo-v2

alannnc opened the pull request:
## What does this PR do?

- Removes old feature that triggered a profile forwarding.
- New feature allows to view current user calendar with out of office days distinction and a Click to Action if forwarded.
- Revamps Out of office from settings
- Adds reasons and a emoji to out of office entries
- New reasons table with seeded for default options

Fixes #13558

 Loom Video: https://www.loom.com/share/2c906d2c73ba4e4aabee4cc77b6b3d23?sid=5327dc00-c8c2-40ae-98e6-ac96f8e4c9fc


## Requirement/Documentation

<!-- Please provide all documents that are important to understand the reason of that PR. -->

- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

## Type of change

<!-- Please delete bullets that are not relevant. -->

- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

- Are there environment variables that should be set? None
- What are the minimal test data to have? 2 users in a team with events
- What is expected (happy path) to have (input and output)? One user can mark days OOO with/without forward to team member.
- Any other important info that could help to test that PR

## Mandatory Tasks

- [X] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

<!-- Remove bullet points below that don't apply to you -->

alannnc committed:
first commit, changes in getSlots endpoint to deliver ooo data

alannnc committed:
Removing old features handle type and user

alannnc committed:
merge with main

alannnc committed:
remove unnecesary import

alannnc committed:
Remove old files from away or booking redirection

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
wip

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
merge with remote

alannnc committed:
wip

alannnc committed:
changed trpc response type and removing no forward ooo days from availability, ui week month column done

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
redo of out of office page, still wip

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
reasons in ooo entry

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
clean up and add missing translations

alannnc committed:
fix text translation components

alannnc committed:
solve merge with remote

alannnc committed:
Add reasons table and default options and implement it on booker

alannnc commented on apps/web/modules/users/views/users-public-view.getServerSideProps.tsx:
```diff
@@ -3,7 +3,6 @@ import type { GetServerSideProps } from "next";
 import { encode } from "querystring";
 import type { z } from "zod";
 
-import { handleUserRedirection } from "@calcom/features/booking-redirect/handle-user";
```
Changes on this file are to remove previous feature of profile forwarding, no longer needed.

alannnc commented on apps/web/modules/users/views/users-public-view.tsx:
```diff
@@ -88,25 +88,6 @@ export function UserPage(props: InferGetServerSidePropsType<typeof getServerSide
-                😄
-              </p>
-            </div>
-          )}
```
No longer needed.

alannnc commented on packages/core/getUserAvailability.ts:
```diff
@@ -343,7 +344,26 @@ const _getUserAvailability = async function getUsersWorkingHoursLifeTheUniverseA
+    availability,
+  });
+
+  // If the user is OOO with a forward to a member that day should be displayed as available
```
My overall take it's deliver the right data from the backend and avoid post processing on the frontend.

alannnc replied:
In this file we fetch out of office entries from the DB and parse them accordingly so slots and availability functions can use them normally.

alannnc commented on packages/core/getUserAvailability.ts:
```diff
@@ -626,3 +647,150 @@ const _getBusyTimesFromDurationLimits = async (
+    const [flattenDays] = availability.map((a) => ("days" in a ? a.days : []));
+    const endDateRange = dayjs(end).utc().endOf("day");
+
+    // Return date is kinda tricky to obtain but basically is the first available date after the end date
```
Return date is no longer needed in the UI

alannnc commented on packages/features/bookings/Booker/Booker.tsx:
```diff
@@ -256,26 +251,6 @@ const BookerComponent = ({
           layout === BookerLayouts.MONTH_VIEW ? "overflow-visible" : "overflow-clip"
         )}>
-        {/* redirect from other user profile */}
-        {isRedirect && (
```
No longer required.

alannnc commented on packages/features/bookings/Booker/components/LargeCalendar.tsx:
```diff
@@ -37,10 +37,14 @@ export const LargeCalendar = ({
-        end: dayjs(slot.time).add(eventDuration, "minutes").toDate(),
-      }));
+      availableTimeslots[day] = schedule.slots[day].map((slot) => {
+        const { time, ...rest } = slot;
```
Here I'm allowing more props to be passed in slot. This should work for a long run since we now have defined our UI implementations. Later down we can explore app dir approach for this component and be served with what the UI need directly from the server.

alannnc commented on packages/features/bookings/Booker/utils/event.ts:
```diff
@@ -13,7 +13,7 @@ export type useScheduleForEventReturnType = ReturnType<typeof useScheduleForEven
 /**
  * Wrapper hook around the trpc query that fetches
- * the event curently viewed in the booker. It will get
+ * the event currently viewed in the booker. It will get
```
typo

alannnc commented on packages/features/bookings/components/AvailableTimes.tsx:
```diff
@@ -162,7 +166,7 @@ const SlotItem = ({
             </HoverCard.Trigger>
             <HoverCard.Portal>
               <HoverCard.Content side="top" align="end" sideOffset={2}>
-                <div className="text-emphasis bg-inverted text-inverted w-[var(--booker-timeslots-width)] rounded-md p-3">
```
text-inverted is the same as text-emphasis

alannnc commented on packages/features/bookings/components/AvailableTimes.tsx:
```diff
@@ -190,6 +194,19 @@ export const AvailableTimes = ({
   const { t } = useLocale();
 
+  if (slots[0] && slots[0].away) {
+    const { toUser, fromUser, reason, emoji } = slots[0];
```
Receiving new props in slots and returning UI early for displaying new UI that allows to switch to different profile.

alannnc commented on packages/features/calendars/weeklyview/components/Calendar.tsx:
```diff
@@ -42,10 +42,10 @@ export function Calendar(props: CalendarComponentProps) {
   const hourSize = 58;
 
-  // Initalise State on inital mount
+  // Initalise State on initial mount
```
typo

alannnc commented on packages/features/calendars/weeklyview/components/Calendar.tsx:
```diff
@@ -81,7 +81,7 @@ export function Calendar(props: CalendarComponentProps) {
                   containerOffsetRef={containerOffset}
                 />
-                <VeritcalLines days={days} />
+                <VerticalLines days={days} />
```
typo

alannnc commented on packages/features/calendars/weeklyview/components/event/Empty.tsx:
```diff
@@ -26,9 +27,9 @@ export function EmptyCell(props: EmptyCellProps) {
     timezone: props.timezone,
   });
 
-  const minuesFromStart = (cellToDate.hour() - props.startHour) * 60 + cellToDate.minute();
```
typo

alannnc commented on packages/features/schedules/lib/use-schedule/useSlotsForDate.ts:
```diff
@@ -3,7 +3,7 @@ import { useMemo } from "react";
 
 /**
- * Get's slots for a specific date from the schedul cache.
+ * Get's slots for a specific date from the schedule cache.
```
typo

alannnc committed:
fixes

alannnc committed:
Merge branch 'main' into ooo-v2

alannnc committed:
return away from ssr

alannnc committed:
Merge branch 'ooo-v2' of github.com:alannnc/cal.com into ooo-v2

alannnc committed:
return away from ssr

alannnc committed:
fix type in test

alannnc commented:
@CarinaWolli 
Marked as ready to review, a few things missing like update on the previous test and handling orgs usernames.
I tried playing with orgs in my local development and made it work, just a few concerns that are not so clear to me right now that are only the org owner gets a special url for his profile while other members get base url.
This means org admin gets -> acme.localhost:3000/owner1/event
While other members get -> localhost:3000/memberUsername
Probably I'm missing some env values here, but with your help I can manage this special case.

Overall I think with this changes we can set a base for the next changes on the org Out of office, but I don't want to go that far without a review here.

zomars committed:
Merge branch 'main' into pr/13621

alannnc committed:
Merge branch 'ooo-v2' of github.com:alannnc/cal.com into ooo-v2

PeerRich commented:
should we also be able to leave a custom comment optionally for internal analytics

alannnc commented:
@PeerRich Sure will do, both of those are easy. Great

alannnc commented:
<img width="638" alt="image" src="https://github.com/calcom/cal.com/assets/4461358/7322455f-9f6d-405e-a659-445d2455ccb0">

- Added query param it is `om` with any value should be truthy (om - open modal)
- Added notes extra field
- Bg color for box with bg-muted


@PeerRich

alannnc committed:
bg-muted + new field notes ooo + query param to open modal

CarinaWolli commented:
> @CarinaWolli Marked as ready to review, a few things missing like update on the previous test and handling orgs usernames. I tried playing with orgs in my local development and made it work, just a few concerns that are not so clear to me right now that are only the org owner gets a special url for his profile while other members get base url. This means org admin gets -> acme.localhost:3000/owner1/event While other members get -> localhost:3000/memberUsername Probably I'm missing some env values here, but with your help I can manage this special case.
> 
> Overall I think with this changes we can set a base for the next changes on the org Out of office, but I don't want to go that far without a review here.

All org members should have  acme.localhost:3000
Also, I noticed that the org links aren't working for me at all in your branch (`http://acme.cal.local:3000/owner1/owner1-acme-event` and `http://acme.cal.local:3000/owner1` shows 404), I don't have any issues with that in main

CarinaWolli reviewed (dismissed):
GreatI  work @alannnc, I am still missing some things to review like slots and availability logic but I have some first feedback for you. 

- We need to find a better way to show this OOO block, maybe start and end it where the working hours start and end would be? (if that's not enough space just expand below?). Also, we should always be able to see the week day, right now this is hidden under the OOO block

![Screenshot 2024-03-12 at 1 54 25 PM](https://github.com/calcom/cal.com/assets/30310907/cacc33bf-9160-4fe9-bf50-42f112c209eb)

CarinaWolli commented:
Also e2e tests needs to be fixed, but I think you are already aware of that 🙏

alannnc commented on packages/prisma/schema.prisma:
```diff
@@ -1147,6 +1152,17 @@ model OutOfOfficeEntry {
   @@index([start, end])
 }
 
+model OutOfOfficeReason {
```
Yeah my initial thought too, but seeing designs for upcoming work here I think we better set on this than later.

alannnc commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -148,25 +143,109 @@ const OutOfOfficeSection = () => {
+            {/* Notes input */}
+            <div className="mt-4">
+              <p className="text-emphasis block text-sm font-medium">{t("notes")}</p>
+              <TextArea
```
Recently implemented it by Peer request.
@PeerRich Wyt about displaying it there for now?

alannnc replied:
Will add them there, sure.

alannnc committed:
minor text fixes anc tw class

alannnc committed:
Fix styles and tw classes

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc committed:
fix orgs getting url params in app router

alannnc committed:
remove org from booker store as not needed

alannnc committed:
redo test for ooo

alannnc committed:
Merge branch 'main' into ooo-v2

alannnc commented:
New updates:

- redo the e2e test for the ooo feature.
- added notes field to display on entries list
- fixed z-index of OOOCard on /booking
- fixed height of OOOCard to be similar to availability from other days, by adding a new slot to that date with { start: endRange }, so I can calculate height based on date diff.
- found main root orgs error, was due [this bug](https://github.com/calcom/cal.com/pull/14076)
- overall I think if we are in orgs url, router.push should jump to teammate username which should be picked up in orgsDomain too so that part we should be fine.

@CarinaWolli

CarinaWolli reviewed (commented):
Finished my second round of review, maybe @emrysal could also take a look at all changes related to slots and availability logic

CarinaWolli commented on packages/core/getUserAvailability.ts:
```diff
@@ -626,3 +647,143 @@ const _getBusyTimesFromDurationLimits = async (
+      OR: [
+        // outside of range
+        // (start <= 'dateTo' AND end >= 'dateFrom')
+        {
+          start: {
+            lte: dateTo.toISOString(),
+          },
+          end: {
+            gte: dateFrom.toISOString(),
+          },
+        },
```
Shouldn't this alone already cover everything?

alannnc replied:
I identified 3 cases, for a ooo range:
- Can be outside of those two dates but have dates in between.
- Can be between two dates (startMonth, endMonth)
- ooo.rangeStart is before startMonth but ooo.rangeEnd is before endMonth

PeerRich commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -148,25 +143,109 @@ const OutOfOfficeSection = () => {
+        <DialogFooter showDivider noSticky>
+          <div className="flex">
+            <Button color="minimal" type="button" onClick={() => closeModal()} className="mr-1">
+              Cancel
```
```suggestion
              {t("cancel")}
```

PeerRich replied:
```suggestion
              {t("cancel")}
```

PeerRich committed:
Update apps/web/pages/settings/my-account/out-of-office/index.tsx

PeerRich commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -178,31 +257,43 @@ const OutOfOfficeEntriesList = () => {
     },
   });
-  if (data === null || data?.length === 0 || data === undefined) return null;
+  if (data === null || data?.length === 0 || (data === undefined && !isPending)) return <EmptyOOO />;
```
yes we should

alannnc committed:
merge with remote

alannnc commented on packages/features/bookings/Booker/components/OutOfOfficeInSlots.tsx:
```diff
@@ -0,0 +1,79 @@
+
+          {fromUser?.displayName && toUser?.displayName && (
+            <p className="text-center text-sm">
+              <Trans
```
Nice catch, this was a side effect from the calendar css. Solved

alannnc committed:
attack review changes, styles for ooo

alannnc committed:
Merge branch 'ooo-v2' of github.com:alannnc/cal.com into ooo-v2

alannnc committed:
merge with remote

alannnc committed:
fixing timezone issues with ooo slots

alannnc committed:
merge with main

alannnc committed:
improve performance to calculate slots for empty column_view

alannnc committed:
fix types

alannnc committed:
fix types

alannnc committed:
fix types

Commit:
Merge branch 'main' into ooo-v2

alannnc committed:
Merge branch 'main' into ooo-v2

CarinaWolli commented:
Can we remove the OOO in the column view when there is no forwarding?
<img width="1094" alt="Screenshot 2024-03-22 at 8 08 24 PM" src="https://github.com/calcom/cal.com/assets/30310907/481b8bf4-51a7-4a00-bc9b-cfe08accb80d">

CarinaWolli commented:
And here too, let's not show that: 
<img width="562" alt="Screenshot 2024-03-22 at 8 30 50 PM" src="https://github.com/calcom/cal.com/assets/30310907/0146f6d1-9928-45b7-8685-56fa991ec6ed">

alannnc committed:
Merge branch 'main' into ooo-v2

alannnc committed:
Merge branch 'ooo-v2' of github.com:alannnc/cal.com into ooo-v2

alannnc commented:
Woop something broke but I think was already being taken care of. Let me double check that validation.
@CarinaWolli

alannnc committed:
fix ooo without to user

alannnc committed:
fix ooo in slots component render conditionals

alannnc commented:
@CarinaWolli Should be solved now: 😅
https://github.com/calcom/cal.com/pull/13621/commits/9ae36193f1cf7c8789aa585957d92e146a452d9c

CarinaWolli commented:
Some more comments, mainly UI related:

1. Now the days are blank, days shouldn't show up at all:
<img width="907" alt="Screenshot 2024-03-22 at 9 41 18 PM" src="https://github.com/calcom/cal.com/assets/30310907/270ab736-31ef-43ac-9e58-be9ef7ab01cc">

2. OOO for one day is not working for me, when I put from April 4 - April 4 it shows as April 3 - April 4

3. Is it possible to put the OOO UI to were the actual start and endtime of availability is (earliest availability of the day to latest). I thought that was already fixed 🤔
<img width="641" alt="Screenshot 2024-03-22 at 9 47 39 PM" src="https://github.com/calcom/cal.com/assets/30310907/ce3a4278-b547-4986-b36b-a2da68c65cc9">

4. NIT: Can we make bottom margin same as on the right and left of the dotted line? 
<img width="319" alt="Screenshot 2024-03-22 at 9 53 38 PM" src="https://github.com/calcom/cal.com/assets/30310907/3f3ae568-b6f8-4348-a0a6-952fcfac1820">

CarinaWolli commented:
Also this is a PR is marked as high-risk, so we need somebody of @calcom/foundation to look at this one as well 🙏

alannnc committed:
fix offset top for ooo in slots

alannnc committed:
filter out ooo days in column view with no forwarding

alannnc committed:
add more margin to ooo in slots

alannnc commented:
Fixed 1, 2 and 4.
#3 Is working for me, if you could share more setup details.

https://www.loom.com/share/ef1fafb71aec4baaa71eb3675a2685c0?sid=7f1f0a66-3e90-4cde-bf3b-964c53f82941
@CarinaWolli

CarinaWolli commented:
When I change my working hours after I already have an OOO created, OOO stops working and is shown as available: [OOO - availability bug ](https://www.loom.com/share/73bbc811d375423388e539a14c4f9d3b)


> 3. Is it possible to put the OOO UI to were the actual start and endtime of availability is (earliest availability of the day to latest). I thought that was already fixed 🤔

Looks good now for me too and can't reproduce it no more

alannnc committed:
fix flatten function on irregular availability

alannnc committed:
Merge branch 'main' of github.com:calcom/cal.com into ooo-v2

alannnc commented:
> When I change my working hours after I already have an OOO created, OOO stops working and is shown as available: [OOO - availability bug ](https://www.loom.com/share/73bbc811d375423388e539a14c4f9d3b)
> 

Good one, found the issue on the flatten function on availability. This was due irregular availability but now my flatten function on days should be working fine.

https://github.com/calcom/cal.com/pull/13621/commits/c04f8ec3918c6eb6b1586bfab9e91ca0d6e5a5bc

emrysal commented:
Looks like there's a related type error 🙏

emrysal committed:
Merge branch 'main' into ooo-v2

alannnc committed:
fix types

alannnc committed:
Merge branch 'ooo-v2' of github.com:alannnc/cal.com into ooo-v2

alannnc committed:
Merge branch 'main' into ooo-v2

alannnc commented:
Note: make sure tests don't behave like #14286

alannnc committed:
Merge branch 'main' into ooo-v2

zomars committed:
Merge branch 'main' into pr/13621

Signed-off-by: zomars <zomars@me.com>

# Conflicts:
#	apps/web/pages/settings/my-account/out-of-office/index.tsx
#	apps/web/public/static/locales/en/common.json
#	packages/features/bookings/components/AvailableTimes.tsx
#	packages/ui/components/empty-screen/EmptyScreen.tsx

zomars committed:
Update index.tsx

zomars reviewed (dismissed):
Code LGTM and tested locally. Waiting for checks to pass.

![LGTM](https://media3.giphy.com/media/THWZYFFJueWv6/200.gif)

alannnc committed:
Merge branch 'main' into ooo-v2

PeerRich committed:
Merge branch 'main' into ooo-v2

PeerRich committed:
Merge branch 'main' into ooo-v2

CarinaWolli reviewed (dismissed):
> When I change my working hours after I already have an OOO created, OOO stops working and is shown as available: [OOO - availability bug](https://www.loom.com/share/73bbc811d375423388e539a14c4f9d3b)

Fixed with https://github.com/calcom/cal.com/commit/c04f8ec3918c6eb6b1586bfab9e91ca0d6e5a5bc, tested an works now 🙏

CarinaWolli committed:
Merge branch 'main' into ooo-v2

CarinaWolli committed:
Merge branch 'main' into ooo-v2

CarinaWolli committed:
Merge branch 'main' into ooo-v2

PeerRich committed:
Merge branch 'main' into ooo-v2

zomars committed:
Merge branch 'main' into ooo-v2

CarinaWolli commented:
@zomars do you know why e2e tests keep failing?

zomars commented:
> @zomars do you know why e2e tests keep failing?

Currently investigating. It is consistently failing but I don't see an obvious cause. I don't think it's directly related to this PR. which is odd.

alannnc commented:
> > @zomars do you know why e2e tests keep failing?
> 
> Currently investigating. It is consistently failing but I don't see an obvious cause. I don't think it's directly related to this PR. which is odd.

It's happening on other PRs aswell? Do we have artifacts to look at? @zomars

zomars commented:
> > > @zomars do you know why e2e tests keep failing?
> > 
> > 
> > Currently investigating. It is consistently failing but I don't see an obvious cause. I don't think it's directly related to this PR. which is odd.
> 
> It's happening on other PRs as well? Do we have artifacts to look at? @zomars

You should be able to download them at the bottom of the results page:

https://github.com/calcom/cal.com/actions/runs/8621332198?pr=13621

zomars commented:
Found the issue. Will PR a patch soon.

zomars committed:
Update OutOfOfficeInSlots.tsx

alannnc committed:
Merge branch 'main' into ooo-v2
